### ISSUE TITLE
bridge: Allow configuring port isolation

### ISSUE THREAD

**Issue by eirikrye** [containers/netavark#1009]
Hello,

I am trying to create podman networks where containers cannot:

1. communicate with each other inside their own bridge (L2)
2. communicate with containers on other bridges (L3)

Setting the `bridge` option `isolate: true` solves the second item. However, containers on the same bridge are still able to communicate.

For the first item, I am able to disallow L2 communication by setting the `isolated on` option on all the bridge ports manually, e.g.:

```
bridge link set dev veth0 isolated on
bridge link set dev veth1 isolated on
```

Is there a way to do this automatically, with netavark, as the bridge ports are created? Alternatively, if I am approaching this issue from the wrong end, is there a better way to achieve what I am looking for?

**Comment by Luap99** [comment]
I think docker calls this inter container connectivity (icc) so this is definitely something we want to support in order to allow better compatibility.

There isn't really anything pluggable which would allow you to set this automatically right now. So this would need to be implemented first. My thinking is to add a new `icc` option and the set the proper netlink attribute to block the connectivity between containers. PRs welcome.

cc @mheon

**Comment by eirikrye** [comment]
Thanks for the response!

Is there a way to hook into the network creation lifecycle with a shell script or something similar?

I quickly tried looking at the plugin API, and tried putting a tiny shell script in `/usr/local/lib/netavark` that just dumps stdin to a file, but I'm not seeing this script being run after setting `netavark_plugin_dirs = ["/usr/local/lib/netavark"]` in `/etc/containers/containers.conf`.

**Comment by Luap99** [comment]
netavark plugin are specified in https://github.com/containers/netavark/blob/main/plugin-API.md but this isn't really what you want. You would need to completely reimplement the entire bridge code basically and then add your extra change.

There are oci hooks which would be more what you are looking for I think but there you have no relation between interface <-> container so you do not know which veth interface to pick.

That is why I said it is a new feature that has to be implemented first.

**Comment by mheon** [comment]
Concur with @Luap99 - seems like an eminently reasonable feature request, and not hard to implement, but will have to be in the existing bridge code.

**Comment by chetan-reddy** [comment]
@eirikrye Were you able to get behavior equivalent to `com.docker.network.bridge.enable_icc` using either a netavark plugin or an OCI hook? If yes, could you please share your code?

**Comment by chetan-reddy** [comment]
Adding the following to my unit files does the job for me. 

`ExecStartPost=bash -c 'bridge link show master bridgename | cut -d: -f2 | cut -d@ -f1 | xargs -t -i bridge link set dev {} isolated on'`

You can set the bridge name using the `--interface-name` option of `podman network create`

**Comment by shivkr6** [comment]
I'm working on this issue.

**Comment by shivkr6** [comment]
Hey @Luap99 @mheon , the netlink attribute required for setting `IFLA_BRPORT_ISOLATED = 1` is not implemented yet in netlink-packet-route: https://docs.rs/netlink-packet-route/latest/netlink_packet_route/link/enum.LinkProtoInfoBridge.html 

Basically what we have to do is:
1) Header
- nlmsg_type=RTM_SETLINK
Use RouteNetlinkMessage::SetLink(msg)

- nlmsg_flags=NLM_F_ACK | NLM_F_REQUEST
- Set the msg.header.index in LinkMessage to the ifi index we get using get_link().
- msg.header.interface_family = AddressFamily::Bridge;

2) Payload (nested)

- IFLA_PROTINFO: this is where bridge-specific settings go. Since veth is a bridge port, this is relevant.
netlink_packet_route::link::LinkProtoInfoBridge

  - IFLA_BRPORT_ISOLATED = 1 : "isolated" flag on a bridge port.
  Needs to be implemented in netlink-packet-route

NOTE: IFLA_BRPORT_ISOLATED should be nested inside IFLA_PROTINFO.

Should I go ahead and implement this in https://github.com/rust-netlink/netlink-packet-route ?

**Comment by Luap99** [comment]
Yes if the library does not parse that attribute yet it should be added there as first step.
